ecosystem/op-deployer.md

```diff
@@ -0,0 +1,163 @@
+fundDevAccounts = true
+# Define the version of the smart contracts to deploy
+contractsVersion = "1.2.0"
+# Override specific variables in the deploy config globally
+overrides = { }
+
+# Define the chains to deploy, by chain ID
+[intent.chains.901]
+# Override variables in the deploy config for this specific chain
+overrides = { }
+
+[intent.chains.902]
+# ... etc.
+```
+
+Under the hood, `op-deployer` will run through the deployment pipeline. The output of each intermediate stage will
+be stored by default within the intent's TOML file, however the implementation of the "state store" will be
+pluggable in order to support use cases like `opc` which store the deployment state in a database. Example output from a
+stage is below:
+
+```toml
+[intent]
+# Intent data is same as above and elided
+
+[state.chains.901]
+# Ownership addresses derived from a mnemonic
+proxyAdminOwner = "0xe59a881b2626f948f56f509f180c32428585629a"
+finalSystemOwner = "0xb9cdf788704088a4c0191d045c151fcbe2db14a4"
+baseFeeVaultRecipient = "0xbc4a9110dad00b4d9fb61743598848ddda6eeb03"
+l1FeeVaultRecipient = "0x6b0c2542fa2cadced5c7f64ef6fb9ebbce7630ff"
+sequencerFeeVaultRecipient = "0xb4e5b724bbc54c95d292613d956871281120ead6"
+
+# Address of OPSM so that other addresses can be retrieved from on-chain data
+opStackManagerAddress = "0x79c6c6b1844e3db7c30107f189cfb095bd2c4b5d"
+
+# Genesis data
+genesis = "base64://abcd..."
+genesisHash = "0x1234..."
+genesisTime = 1234567890
+
+[state.chains.902]
+# ... etc.
+```
+
+Only the minimum necessary state is stored. For example, the rollup config can be derived from deploy config
+variables and therefore can be generated on-the-fly when it is needed. `op-deployer` will expose utility subcommands
+output any generated data, such as the rollup config, to the console.
+
+`op-deployer` itself will be a single Go binary that can be run from the command line. Some stages - like those that
+interact with Forge tooling to deploy contracts - will either shell out to other tools or orchestrate Docker
+containers that run the necessary tools. This way both local development and pre-packaged use cases are supported.
+Given the overall migration towards Go tooling over Forge for genesis creation and deployments, we expect that the
+number of tools requiring Docker/shell integrations will decrease over time.
+
+Each stage will be represented as a standalone Go function for easy integration with additional tooling.
+
+## Deploy Config Cleanup
+
+The deployment config file is a hodgepodge of almost 100 variables which all need to be set in order to deploy a chain.
+Some of these variables are only used in specific contexts, but all of them need to be set in order for some of our
+tools to work. Additionally, the deploy config expects to deploy a single L2, which makes it impossible to deploy
+multiple chains.
+
+To fix this, we will:
+
+- Remove the L1 genesis variables from the deploy config. Deploying a new l1 should be a separate process from
```

Comment:
@protolambda split up the deploy config into themes already but did it in a backwards compatible way so that there are no breaking changes to the JSON. Would it be useful to have a migration tool, like input legacy deploy config and output a new split up deploy config?

Reply:
I can certainly do this, but IMO we should just migrate everything all at once to the new format.

Reply:
I did, but realistically the deploy config still has tech-debt, in naming of attributes and other ways. I would center the op-deployer around OPSM configs